kube-runtime/src/shared_store/safe_store.rs

```diff
@@ -0,0 +1,96 @@
+use crate::reflector::{ObjectRef, Store};
+use crate::shared_store::ready_token::ReadyToken;
+use kube_client::Resource;
+use std::hash::Hash;
+use std::sync::Arc;
+
+/// A wrapper around a Store that exposes async versions of the Store's methods. These will always
+/// return immediately if the store is ready, or wait until the store is ready if it is not.
+#[derive(Clone)]
+pub struct SafeStore<K: 'static + Resource>
+where
+    K::DynamicType: Hash + Eq,
+{
+    store: Store<K>,
+    ready: ReadyToken,
+}
+
+impl<K: 'static + Clone + Resource> SafeStore<K>
+where
+    K::DynamicType: Eq + Hash + Clone,
+{
+    pub fn new(store: Store<K>) -> SafeStore<K> {
+        Self {
+            store,
+            ready: ReadyToken::new(),
+        }
+    }
+
+    pub fn make_ready(&self) {
+        self.ready.make_ready()
+    }
+
+    pub async fn get(&self, key: &ObjectRef<K>) -> Option<Arc<K>> {
+        self.ready.ready().await;
+        self.store.get(key)
+    }
```

Comment:
I am a little worried about everything hinging on this future not being ready. This means we can pass around SafeStores that are not actually ready. Given that "being ready" is a one-time thing to wait, it feels to me that this should be done as we transition INTO the SafeStore. This also would simplify our usage in more complicated bits because we can just assume the invariant than have to account for the possibility of it with async (which is sometimes hard inside combinator style async code).

Reply:
effectively what i want is an async TryFrom style conversion (maybe via https://docs.rs/async-convert/latest/async_convert/trait.TryFrom.html - haven't really evaluated this crate)

Reply:
I think the main thing I'm trying to address is decoupling setup from usage. If the await was on the transition into of `SafeStore` you still need to start consuming the watcher stream before you could get a `SafeStore` to pass to a `Controller`. That would make wiring together lots of stores and controllers much more complicated. 

Maybe a better alternative would be to return a `Result::Err` from the store if not ready; at least that way if you do hit the race condition (or you forget to start consuming your watcher stream) you just get an error `StoreNotReady` rather than an `await` that could potentially hang forever.

I'll focus on tidying up the `StreamSubscribe` for now then circle back around to think more about this (and have a look into the crate you've linked).